### PR TITLE
Support name tokens

### PR DESCRIPTION
**Pull request by bjlittle**
This PR is the outcome of a side discussion regarding issue https://github.com/SciTools/iris/issues/3360.

In short, in `iris` we've allowed the use of `long_name` from a coordinate to be used in the construction of the `CellMethod`, but that causes issues when the `long_name` contains one or more space (" ") characters, as it then quickly becomes ambiguous to parse the grammar suggested in the CF Conventions for a `cell_methods` attribute on loading after saving the `CellMethod` to NetCDF. That's not a fault of the CF Conventions, more our interpretation and implementation of it (from a file format agnostic point of view).

Reference, [Cell Methods](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#cell-methods), 2nd paragraph:

> In the specification of this attribute, name can be a dimension of the variable, a scalar coordinate variable, a valid standard name, or the word "area"

That said, this PR attempts to resolve this issue by ensuring that the `coord_names` and `method` used within the construction of a `CellMethod` are [valid NetCDF name tokens](https://www.unidata.ucar.edu/software/netcdf/docs/netcdf_data_set_components.html#object_name), and thus can be parsed within the rules suggested in the CF Conventions without ambiguity (or over-engineering :wink:).

### CODE REVIEW

**Comment by bjlittle** [comment]
@lbdreyer I still need to add some test coverage to the PR... just pushed this so that you can see the approach I'm taking :wink:

File: lib/iris/_cube_coord_common.py [hunk 1]
```diff
@@ -19,15 +19,18 @@
 from six.moves import (filter, input, map, range, zip)  # noqa
 import six
 
-# TODO: Is this a mixin or a base class?
-
```
Comment (bjlittle) [inline lib/iris/_cube_coord_common.py]: Horse. Bolted.

**Comment by bjlittle** [comment]
@lbdreyer Okay, I think that I'm pretty much there...

As it happens, there wasn't any unit test coverage of `CellMethods` (there is for parsing, but not construction and usage), but I've added the minimal test coverage for the changes in this PR.

Being a good citizen, I'll add test coverage in a separate PR.

I've also not updated the docs either, just so that this PR can get banked sooner rather than later, but happy to make a follow-up PR, if you think that's appropriate... :smile:

File: lib/iris/tests/unit/cube_coord_common/test_CFVariableMixin.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+    def test_var_name(self):
+        token = 'var_name'
+        self.cf_var.var_name = token
+        result = self.cf_var.name()
+        self.assertEqual(result, token)
+
+    def test_stash(self):
+        token = 'stash'
+        self.cf_var.var_name = token
```
Comment (lbdreyer) [inline lib/iris/tests/unit/cube_coord_common/test_CFVariableMixin.py]: This doesn't test stash, this repeats the var_name test? I'm guessing this was a copy and paste error?

File: lib/iris/tests/unit/cube_coord_common/test_CFVariableMixin.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    def test_token_default(self):
+        result = self.cf_var.name(token=True)
+        self.assertEqual(result, self.default)
+
+    def test_fail_token_default(self):
+        cf_var = CFVariableMixin()
+        cf_var.standard_name = None
+        cf_var.long_name = None
+        cf_var.var_name = None
+        cf_var.attributes = {}
```
Comment (lbdreyer) [inline lib/iris/tests/unit/cube_coord_common/test_CFVariableMixin.py]: Is this necessary? I can't see a difference between `cf_var` and `self.cf_var`

File: lib/iris/_cube_coord_common.py [hunk 4]
```diff
@@ -144,12 +196,10 @@ def var_name(self):
...
-            if not name:
-                raise ValueError('An empty string is not a valid netCDF '
-                                 'variable name.')
-            elif set(name).intersection(string.whitespace):
-                raise ValueError('{!r} is not a valid netCDF variable name '
-                                 'as it contains whitespace.'.format(name))
+            result = self.token(name)
+            if result is None or not name:
+                emsg = '{!r} is not a valid NetCDF variable name.'
+                raise ValueError(emsg.format(name))
```
[the remark this answers is not in the document]
Comment (bjlittle) [reply lib/iris/_cube_coord_common.py]: Not sure either. But let's err on the side of this change upsetting someone... so I'll add an entry into the `whatsnew` 👍

File: lib/iris/tests/unit/cube_coord_common/test_CFVariableMixin.py (hunk 2 above)
Comment (bjlittle) [reply lib/iris/tests/unit/cube_coord_common/test_CFVariableMixin.py]: Yup, good spot 👀

File: lib/iris/tests/unit/cube_coord_common/test_CFVariableMixin.py (hunk 3 above)
Comment (bjlittle) [reply lib/iris/tests/unit/cube_coord_common/test_CFVariableMixin.py]: Yup, I can purge the use of `cf_var`, thanks!

**Comment by bjlittle** [comment]
@lbdreyer I've actioned all your review comments, `travis-ci` is happy, over to you... 😺

**Comment by lbdreyer** [comment]
All good :+1: :+1: Thanks @bjlittle !

**Pull request merged**